src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1211,12 +1239,17 @@
 (pls/defn-validated deactivate-node!
   "Deactivate the given host, recording the current time. If the node is
   currently inactive, no change is made."
-  [certname :- String & [timestamp :- pls/Timestamp]]
-  (let [timestamp (to-timestamp (or timestamp (now)))]
-    (sql/do-prepared "UPDATE certnames SET deactivated = ?
-                        WHERE certname=?
-                           AND (deactivated IS NULL OR deactivated < ?)"
-                     [timestamp certname timestamp])))
+  ([certname :- String]
+   (deactivate-node! certname (now)))
+  ([certname :- String timestamp :- pls/Timestamp]
+   (if (have-newer-record-for-certname? certname timestamp)
+     (log/warnf "Not deactivating node %s because local data is newer than %s."
+                certname timestamp)
+     (let [sql-timestamp (to-timestamp timestamp)]
+       (sql/do-prepared "UPDATE certnames SET deactivated = ?
+                           WHERE certname=?
+                             AND (deactivated IS NULL OR deactivated < ?)"
+                        [sql-timestamp certname sql-timestamp])))))
 
```

Comment:
In addition to the PR-related changes, I broke this up into two arities to handle the default timestamp more directly.